### PR TITLE
Deterministic Geometry Proposal

### PR DESCRIPTION
**Pull request by dominicrufa**
## Description
allow for deterministic placement of `unique_new` atoms by bypassing stochastic placemement in `geometry.py`

## TODO:
- [x] implement changes to `RelativeFEPSetup`
- [x] write test
- [x] assert `core_atoms` are unchanged in position

### CODE REVIEW

**Comment by jchodera** [comment]
@dominicrufa : I've pushed a small test for prespecified geometries:
```
perses/data/given-geometries/
  ligands.sdf - a series of 10 molecules that differ by only a few atoms, with all core atoms sharing the same positions
  receptor.pdb - a 3-residue peptide segment of SARS-CoV-2 Mpro extracted from the cocrystal structure and left zwitterionic at the termini
```
Hopefully all of the ligands should be workable using our current mapping criteria, but we could add a mapping criteria mode that simply uses the atoms that are right on top of each other to derive the mapping if we want to guarantee this.

**Comment by dominicrufa** [comment]
error seems unrelated to the new test.

**Comment by jchodera** [comment]
@dominicrufa : Is this ready to merge?

**Comment by jchodera** [comment]
Can we also cut a new release after this?

File: perses/app/relative_setup.py [hunk 1]
```diff
@@ -158,6 +162,13 @@ def __init__(self,
         self._complex_box_dimensions = complex_box_dimensions
         self._solvent_box_dimensions = solvent_box_dimensions
         self._map_strategy = map_strategy
+        self._use_given_geometries = use_given_geometries
+        self._ligand_input = ligand_input
+
+        if self._use_given_geometries:
+            assert self._ligand_input[-3:] == 'sdf' or self._ligand_input[-4:] == 'mol2', f"cannot use deterministic atom placement if the ligand input files do not contain geometry information (e.g. in .sdf or .mol2 format)"
```
Comment (jchodera) [inline perses/app/relative_setup.py:169]: You can use `self._ligand_input.endsiwth('.sdf')` instead, or even 
```python
import os
root, ext = os.path.splitext(self._ligand_input)
assert ext in set(['sdf', 'SDF', 'mol2'])
```

**Comment by jchodera** [comment]
@dominicrufa : Do you think we'll also need to add an atom-mapping scheme where we derive the atom map from which atoms are right on top of each other?

**Comment by jchodera** [comment]
Awesome, thanks!

I'll take a quick look at the errors too.

File: perses/app/setup_relative_calculation.py [hunk 2]
```diff
@@ -559,9 +570,12 @@ def run_setup(setup_options, serialize_systems=True, build_samplers=True):
             _forward_added_valence_energy = top_prop['%s_added_valence_energy' % phase]
             _reverse_subtracted_valence_energy = top_prop['%s_subtracted_valence_energy' % phase]
 
-            zero_state_error, one_state_error = validate_endstate_energies(_top_prop, _htf, _forward_added_valence_energy, _reverse_subtracted_valence_energy, beta = 1.0/(kB*temperature), ENERGY_THRESHOLD = ENERGY_THRESHOLD)#, trajectory_directory=f'{xml_directory}{phase}')
-            _logger.info(f"\t\terror in zero state: {zero_state_error}")
-            _logger.info(f"\t\terror in one state: {one_state_error}")
+            if not setup_options['use_given_geometries']:
```
Comment (jchodera) [inline perses/app/setup_relative_calculation.py:573]: This won't work, and causes this error:
```
======================================================================
ERROR: perses.tests.test_fah.test_pipeline_small_molecule_solvent
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/share/miniconda/envs/test/lib/python3.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/runner/work/perses/perses/perses/tests/test_fah.py", line 71, in test_pipeline_small_molecule_solvent
    num_equilibration_steps_per_iteration = DEFAULT_EQ_STEPS_PER_ITERATION)
  File "/home/runner/work/perses/perses/perses/app/fah_generator.py", line 359, in run_neq_fah_setup
    setup_dict = run_setup(setup_options, serialize_systems=False, build_samplers=False)
  File "/home/runner/work/perses/perses/perses/app/setup_relative_calculation.py", line 573, in run_setup
    if not setup_options['use_given_geometries']:
KeyError: 'use_given_geometries'
-------------------- >> begin captured stdout << ---------------------
```
Instead, you want to check if the key is present first:
```python
if not 'use_given_geometries' in setup_options:
```

**Comment by zhang-ivy** [comment]
One of the errors we are seeing when I ran `perses-fah` using this branch is:
```2020-10-08 12:54:53,270:(0.00s):relative:Handling nonbonded forces...
2020-10-08 12:54:53,270:(0.00s):relative:	handle_nonbonded: looping through all particles in hybrid...
2020-10-08 12:54:53,546:(0.28s):relative:	handle_nonbonded: Handling Interaction Groups...
2020-10-08 12:54:53,548:(0.00s):relative:	handle_nonbonded: Handling Hybrid Exceptions...
2020-10-08 12:54:53,551:(0.00s):relative:	handle_nonbonded: Handling Original Exceptions...
Traceback (most recent call last):
  File "/home/zhangi/miniconda3/envs/perses-sims/bin/perses-fah", line 11, in <module>
    load_entry_point('perses==0.8.2.dev0', 'console_scripts', 'perses-fah')()
  File "/home/zhangi/miniconda3/envs/perses-sims/lib/python3.7/site-packages/perses-0.8.2.dev0-py3.7.egg/perses/app/fah_generator.py", line 478, in run
    run_neq_fah_setup(**setup_options)
  File "/home/zhangi/miniconda3/envs/perses-sims/lib/python3.7/site-packages/perses-0.8.2.dev0-py3.7.egg/perses/app/fah_generator.py", line 359, in run_neq_fah_setup
    setup_dict = run_setup(setup_options, serialize_systems=False, build_samplers=False)
  File "/home/zhangi/miniconda3/envs/perses-sims/lib/python3.7/site-packages/perses-0.8.2.dev0-py3.7.egg/perses/app/setup_relative_calculation.py", line 559, in run_setup
    interpolate_old_and_new_14s = setup_options['anneal_1,4s'])
  File "/home/zhangi/miniconda3/envs/perses-sims/lib/python3.7/site-packages/perses-0.8.2.dev0-py3.7.egg/perses/annihilation/relative.py", line 300, in __init__
    self.handle_nonbonded()
  File "/home/zhangi/miniconda3/envs/perses-sims/lib/python3.7/site-packages/perses-0.8.2.dev0-py3.7.egg/perses/annihilation/relative.py", line 1424, in handle_nonbonded
    self._handle_original_exceptions()
  File "/home/zhangi/miniconda3/envs/perses-sims/lib/python3.7/site-packages/perses-0.8.2.dev0-py3.7.egg/perses/annihilation/relative.py", line 1713, in _handle_original_exceptions
    [charge1_old, sigma1_old, epsilon1_old] = old_system_nonbonded_force.getParticleParameters(index1_old)
  File "/home/zhangi/miniconda3/envs/perses-sims/lib/python3.7/site-packages/simtk/openmm/openmm.py", line 19448, in getParticleParameters
    val = _openmm.NonbondedForce_getParticleParameters(self, index)
TypeError: in method 'NonbondedForce_getParticleParameters', argument 2 of type 'int'
```
After some sleuthing, I figured out it's because the indices in the map created [here](https://github.com/choderalab/perses/blob/master/perses/app/relative_setup.py#L614-L615): are `numpy int64`, not `int` -- likely due to use of mdtraj to retrieve indices

**Comment by jchodera** [comment]
The Python wrapper for OpenMM is not so tolerant about other integer types. You just have to explicitly cast to `int` with `int(value)`.

Can you also open an OpenMM issue to report that this method needs to be updated to handle `numpy.int64` types?

**Comment by dominicrufa** [comment]
> One of the errors we are seeing when I ran `perses-fah` using this branch is:
> 
> ```
> 2020-10-08 12:54:53,270:(0.00s):relative:	handle_nonbonded: looping through all particles in hybrid...
> 2020-10-08 12:54:53,546:(0.28s):relative:	handle_nonbonded: Handling Interaction Groups...
> 2020-10-08 12:54:53,548:(0.00s):relative:	handle_nonbonded: Handling Hybrid Exceptions...
> 2020-10-08 12:54:53,551:(0.00s):relative:	handle_nonbonded: Handling Original Exceptions...
> Traceback (most recent call last):
>   File "/home/zhangi/miniconda3/envs/perses-sims/bin/perses-fah", line 11, in <module>
>     load_entry_point('perses==0.8.2.dev0', 'console_scripts', 'perses-fah')()
>   File "/home/zhangi/miniconda3/envs/perses-sims/lib/python3.7/site-packages/perses-0.8.2.dev0-py3.7.egg/perses/app/fah_generator.py", line 478, in run
>     run_neq_fah_setup(**setup_options)
>   File "/home/zhangi/miniconda3/envs/perses-sims/lib/python3.7/site-packages/perses-0.8.2.dev0-py3.7.egg/perses/app/fah_generator.py", line 359, in run_neq_fah_setup
>     setup_dict = run_setup(setup_options, serialize_systems=False, build_samplers=False)
>   File "/home/zhangi/miniconda3/envs/perses-sims/lib/python3.7/site-packages/perses-0.8.2.dev0-py3.7.egg/perses/app/setup_relative_calculation.py", line 559, in run_setup
>     interpolate_old_and_new_14s = setup_options['anneal_1,4s'])
>   File "/home/zhangi/miniconda3/envs/perses-sims/lib/python3.7/site-packages/perses-0.8.2.dev0-py3.7.egg/perses/annihilation/relative.py", line 300, in __init__
>     self.handle_nonbonded()
>   File "/home/zhangi/miniconda3/envs/perses-sims/lib/python3.7/site-packages/perses-0.8.2.dev0-py3.7.egg/perses/annihilation/relative.py", line 1424, in handle_nonbonded
>     self._handle_original_exceptions()
>   File "/home/zhangi/miniconda3/envs/perses-sims/lib/python3.7/site-packages/perses-0.8.2.dev0-py3.7.egg/perses/annihilation/relative.py", line 1713, in _handle_original_exceptions
>     [charge1_old, sigma1_old, epsilon1_old] = old_system_nonbonded_force.getParticleParameters(index1_old)
>   File "/home/zhangi/miniconda3/envs/perses-sims/lib/python3.7/site-packages/simtk/openmm/openmm.py", line 19448, in getParticleParameters
>     val = _openmm.NonbondedForce_getParticleParameters(self, index)
> TypeError: in method 'NonbondedForce_getParticleParameters', argument 2 of type 'int'
> ```
> 
> After some sleuthing, I figured out it's because the indices in the map created [here](https://github.com/choderalab/perses/blob/master/perses/app/relative_setup.py#L614-L615): are `numpy int64`, not `int` -- likely due to use of mdtraj to retrieve indices

that means that the complex phase setup works, which is confusingto me since there are solvent particles here, too. I'll look again

**Comment by dominicrufa** [comment]
> 2020-10-13T17:07:43.4789471Z ERROR: execute the `RelativeFEPSetup` with geometries specified for a host-guest relative free energy pair
> 2020-10-13T17:07:43.4791018Z ----------------------------------------------------------------------
> 2020-10-13T17:07:43.4792106Z Traceback (most recent call last):
> 2020-10-13T17:07:43.4793652Z   File "/usr/share/miniconda/envs/test/lib/python3.7/site-packages/nose/case.py", line 197, in runTest
> 2020-10-13T17:07:43.4795396Z     self.test(*self.arg)
> 2020-10-13T17:07:43.4796826Z   File "/home/runner/work/perses/perses/perses/tests/test_relative_setup.py", line 257, in test_host_guest_deterministic_geometries
> 2020-10-13T17:07:43.4798691Z     use_given_geometries = True
> 2020-10-13T17:07:43.4799875Z   File "/home/runner/work/perses/perses/perses/app/relative_setup.py", line 360, in __init__
> 2020-10-13T17:07:43.4802271Z     small_molecule_forcefield=small_molecule_forcefield, molecules=molecules, cache=small_molecule_parameters_cache, periodic_forcefield_kwargs = {'nonbondedMethod': self._nonbonded_method})
> 2020-10-13T17:07:43.4804812Z   File "/usr/share/miniconda/envs/test/lib/python3.7/site-packages/openmmforcefields/generators/system_generators.py", line 211, in __init__
> 2020-10-13T17:07:43.4809883Z     self.add_molecules(molecules)
> 2020-10-13T17:07:43.4811545Z   File "/usr/share/miniconda/envs/test/lib/python3.7/site-packages/openmmforcefields/generators/system_generators.py", line 238, in add_molecules
> 2020-10-13T17:07:43.4812935Z     self.template_generator.add_molecules(molecules)
> 2020-10-13T17:07:43.4814719Z   File "/usr/share/miniconda/envs/test/lib/python3.7/site-packages/openmmforcefields/generators/template_generators.py", line 124, in add_molecules
> 2020-10-13T17:07:43.4816154Z     self._molecules.update( { molecule.to_smiles() : molecule for molecule in molecules } )
> 2020-10-13T17:07:43.4817812Z   File "/usr/share/miniconda/envs/test/lib/python3.7/site-packages/openmmforcefields/generators/template_generators.py", line 124, in <dictcomp>
> 2020-10-13T17:07:43.4819246Z     self._molecules.update( { molecule.to_smiles() : molecule for molecule in molecules } )
> 2020-10-13T17:07:43.4820809Z   File "/usr/share/miniconda/envs/test/lib/python3.7/site-packages/openforcefield/topology/molecule.py", line 1841, in to_smiles
> 2020-10-13T17:07:43.4822183Z     smiles = to_smiles_method(self)
> 2020-10-13T17:07:43.4823582Z   File "/usr/share/miniconda/envs/test/lib/python3.7/site-packages/openforcefield/utils/toolkits.py", line 1032, in to_smiles
> 2020-10-13T17:07:43.4825089Z     oemol = OpenEyeToolkitWrapper.to_openeye(molecule)
> 2020-10-13T17:07:43.4826579Z   File "/usr/share/miniconda/envs/test/lib/python3.7/site-packages/openforcefield/utils/toolkits.py", line 934, in to_openeye
> 2020-10-13T17:07:43.4828361Z     'Programming error: OpenEye atom stereochemistry assumptions failed.'
> 2020-10-13T17:07:43.4829788Z Exception: Programming error: OpenEye atom stereochemistry assumptions failed.

weird. the `.sdf` mols 0 and 1 cannot be created as systems?

**Comment by jchodera** [comment]
> 2020-10-13T17:07:43.4828361Z 'Programming error: OpenEye atom stereochemistry assumptions failed.'
> 2020-10-13T17:07:43.4829788Z Exception: Programming error: OpenEye atom stereochemistry assumptions failed.

That's odd. I thought this was fixed in `openforcefields >=0.7.1`. See [this issue](https://github.com/openmm/openmmforcefields/issues/141) and the linked issue therein.

Which version of the openforcefield toolkit is being installed during unit tests?

**Comment by jchodera** [comment]
@dominicrufa " The tests are picking up `openforcefield` 0.6.0 for some reason:
```
openforcefield            0.6.0                    py37_1    omnia
openforcefields           1.2.1                      py_0    omnia
```

**Comment by jchodera** [comment]
@dominicrufa : Can you pin `openforcefield >=0.7.1` [here](https://github.com/choderalab/perses/blob/68e73977c8677dc9ab655e77fcfab04b7c06ec02/devtools/conda-envs/test_env.yaml#L38)?

**Comment by dominicrufa** [comment]
> @dominicrufa : Can you pin `openforcefield >=0.7.1` [here](https://github.com/choderalab/perses/blob/68e73977c8677dc9ab655e77fcfab04b7c06ec02/devtools/conda-envs/test_env.yaml#L38)?

done.

**Comment by dominicrufa** [comment]
> > @dominicrufa : Can you pin `openforcefield >=0.7.1` [here](https://github.com/choderalab/perses/blob/68e73977c8677dc9ab655e77fcfab04b7c06ec02/devtools/conda-envs/test_env.yaml#L38)?
> 
> done.

though we don't see issues with the jacs set molecules, so there might be something weird with the hostguest ligands? we'll see

**Comment by zhang-ivy** [comment]
> One of the errors we are seeing when I ran `perses-fah` using this branch is:
> 
> ```
> 2020-10-08 12:54:53,270:(0.00s):relative:	handle_nonbonded: looping through all particles in hybrid...
> 2020-10-08 12:54:53,546:(0.28s):relative:	handle_nonbonded: Handling Interaction Groups...
> 2020-10-08 12:54:53,548:(0.00s):relative:	handle_nonbonded: Handling Hybrid Exceptions...
> 2020-10-08 12:54:53,551:(0.00s):relative:	handle_nonbonded: Handling Original Exceptions...
> Traceback (most recent call last):
>   File "/home/zhangi/miniconda3/envs/perses-sims/bin/perses-fah", line 11, in <module>
>     load_entry_point('perses==0.8.2.dev0', 'console_scripts', 'perses-fah')()
>   File "/home/zhangi/miniconda3/envs/perses-sims/lib/python3.7/site-packages/perses-0.8.2.dev0-py3.7.egg/perses/app/fah_generator.py", line 478, in run
>     run_neq_fah_setup(**setup_options)
>   File "/home/zhangi/miniconda3/envs/perses-sims/lib/python3.7/site-packages/perses-0.8.2.dev0-py3.7.egg/perses/app/fah_generator.py", line 359, in run_neq_fah_setup
>     setup_dict = run_setup(setup_options, serialize_systems=False, build_samplers=False)
>   File "/home/zhangi/miniconda3/envs/perses-sims/lib/python3.7/site-packages/perses-0.8.2.dev0-py3.7.egg/perses/app/setup_relative_calculation.py", line 559, in run_setup
>     interpolate_old_and_new_14s = setup_options['anneal_1,4s'])
>   File "/home/zhangi/miniconda3/envs/perses-sims/lib/python3.7/site-packages/perses-0.8.2.dev0-py3.7.egg/perses/annihilation/relative.py", line 300, in __init__
>     self.handle_nonbonded()
>   File "/home/zhangi/miniconda3/envs/perses-sims/lib/python3.7/site-packages/perses-0.8.2.dev0-py3.7.egg/perses/annihilation/relative.py", line 1424, in handle_nonbonded
>     self._handle_original_exceptions()
>   File "/home/zhangi/miniconda3/envs/perses-sims/lib/python3.7/site-packages/perses-0.8.2.dev0-py3.7.egg/perses/annihilation/relative.py", line 1713, in _handle_original_exceptions
>     [charge1_old, sigma1_old, epsilon1_old] = old_system_nonbonded_force.getParticleParameters(index1_old)
>   File "/home/zhangi/miniconda3/envs/perses-sims/lib/python3.7/site-packages/simtk/openmm/openmm.py", line 19448, in getParticleParameters
>     val = _openmm.NonbondedForce_getParticleParameters(self, index)
> TypeError: in method 'NonbondedForce_getParticleParameters', argument 2 of type 'int'
> ```
> 
> After some sleuthing, I figured out it's because the indices in the map created [here](https://github.com/choderalab/perses/blob/master/perses/app/relative_setup.py#L614-L615): are `numpy int64`, not `int` -- likely due to use of mdtraj to retrieve indices

@dominicrufa 's  most recent commit seems to have fixed the bugs in the tests but not the one quoted above.

**Comment by jchodera** [comment]
The OpenMM Python API is picky about wanting `int` instead of `numpy.int64` or other variants, so a simple cast to `int`---as in `int(number)`---will fix this!

@zhang-ivy can you also report this to the OpenMM issue tracker so that we can fix the python wrappers to tolerate this?

**Comment by jchodera** [comment]
Awesome! Tests pass! Is this ready to go?

**Comment by dominicrufa** [comment]
> Awesome! Tests pass! Is this ready to go?

I'm just checking to make sure that the atom map looks correct: 
image
i think this is sufficient

**Comment by jchodera** [comment]
> I'm just checking to make sure that the atom map looks correct:

@dominicrufa : Can you also paste the atom map showing the 2D structures and mapped atoms? 
I

**Comment by dominicrufa** [comment]
> > I'm just checking to make sure that the atom map looks correct:
> 
> @dominicrufa : Can you also paste the atom map showing the 2D structures and mapped atoms?
> I
image
i am using `np.allclose` for x,y,z coords of the atoms as a criterion for mapping. the default tolerances are rather high.
if `atol` is relaxed to 1e-4, then both conjugated rings map.

**Comment by jchodera** [comment]
@dominicrufa : There's something pathologically wrong with that 2D depiction---the hydrogens are all missing! Does this indicate some kind of major bug?

If this is just a visualization issue, it looks like the main problem is that the ring containing atom 12 is rotated, and our mapping scheme is mapping one atom fo the ring but not the rest. I am guessing this would be problematic for perses, so we need to add one more stage that goes through all rings and de-maps all atoms in the ring if any atom in the ring is de-mapped. 

If this isn't a problem for perses, no worries then! It just means that the information about that 11-12 bond vector is preserved between compounds!

**Comment by dominicrufa** [comment]
> @dominicrufa : There's something pathologically wrong with that 2D depiction---the hydrogens are all missing! Does this indicate some kind of major bug?
after some digging, i realized the `.sdf` ligands don't have any of the same positions, so they are not being mapped. 
> 
> If this is just a visualization issue, it looks like the main problem is that the ring containing atom 12 is rotated, and our mapping scheme is mapping one atom fo the ring but not the rest. I am guessing this would be problematic for perses, so we need to add one more stage that goes through all rings and de-maps all atoms in the ring if any atom in the ring is de-mapped.
> 
> If this isn't a problem for perses, no worries then! It just means that the information about that 11-12 bond vector is preserved between compounds!
this isnt a problem. also, since i increased the tolerance of the atom overlap, they are mapped

**Comment by jchodera** [comment]
> after some digging, i realized the .sdf ligands don't have any of the same positions, so they are not being mapped.

Are you referring to the  `ligands.sdf`](perses/data/given-geometries/ligands.sdf) I added in https://github.com/choderalab/perses/pull/740/commits/b4453888dd1fc58c922de711ac7ed48365b44708? 

Here is a superposition of all the ligands in `ligands.sdf`---they're at least close enough visually that they have overlapping atoms. Do you mean that the positions are not exactly identical? If so, could we use a distance threshold of < 0.1 A instead?
ligands

> this isnt a problem. also, since i increased the tolerance of the atom overlap, they are mapped

Ah, great!

Any idea why the protons are being omitted from your 2D mapping depiction? Can we fix that?

**Comment by dominicrufa** [comment]
> Any idea why the protons are being omitted from your 2D mapping depiction? Can we fix that?

that was my post-hoc rendering of the atom map (outside the pipeline). the fah setup generates the proper 2d depiction.
the full `perses-fah` setup seems to run and conduct relaxation for all of the phases, so we are good to go.

**Comment by hannahbrucemacdonald** [comment]
Does the ‘Geometry’ mapping not help with some of these cases? Are you using absolute coordinates to place new atoms, or intramolecular distances/angles/torsions?

**Comment by dominicrufa** [comment]
> Does the ‘Geometry’ mapping not help with some of these cases? Are you using absolute coordinates to place new atoms, or intramolecular distances/angles/torsions?

the motivation here is to use absolute coordinates (within some prespecified tolerance). it seems to map all (presumably) mappable atoms when the tolernance is set to ~0.1Angstrom

**Comment by jchodera** [comment]
@dominicrufa : Is this still "WIP", or is it ready to merge?

**Comment by dominicrufa** [comment]
> @dominicrufa : Is this still "WIP", or is it ready to merge?

this is mergable. will change. feel free to merge/cut release as you see fit.

**Comment by jchodera** [comment]
Thanks!

**Pull request merged**